# Business Analytics - Assignment 2 - Question 3
#### **Student Name:** Koorosh Shakoori

# Chemical Production Composition

In this task we are solving a non-linear Problem using scipy.


In [2]:
import numpy as np
from scipy.optimize import minimize

## Defining the Decision Variables
The decision variables in this problem are the amount to produce of each chemical product 1 and 2.

To define the variables we will make an initial numpy array called **x0** of two elements where:
- Product 1 --> x[0]
- Product 2 --> x[1]

In [3]:
#we will assign an initial guess to each variable
    
x0 = np.zeros(2)
x0[0] = 2.0
x0[1] = 10.0

## Defining the Objective Function
Given the price of the raw material, production cost of each product and the formula for their revenue, the function for the profit of each product can be described as below:
- Product 1 profit: (30 - x[0])\*x[0] - 3\*x[0] - 10\*x[0] --> 17\*x[0] - x[0]\*\*2
- Product 2 profit: (50 - 2\*x[1])\*x[1] - 5\*x[1] - 10\*x[1] --> 35\*x[1] - 2\*x[1]\*\*2

The total profit formula is the summation of both profits above, shown below:
- total profit(to maximize): 17\*x[0] - x[0]\*\*2 + 35\*x[1] - 2\*x[1]\*\*2

since we want to maximize this function and the tools we have allow us only to minimize a function, we will need to transform the formula into a minimization problem by multiplying it with -1:
- total profit(to minimize):  x[0]\*\*2 - 17\*x[0] + 2\*x[1]\*\*2 - 35\*x[1]

In [4]:
#Now we will define the function above into a python function.
def objective(x):
    return x[0]**2 - 17*x[0] + 2*x[1]**2 - 35*x[1]

## Defining the Constraints
According to the problem, only 17.25 ounce of raw material can be acquired. To enforce this we will define a constraint where the summation of both products is less than 17.25.

x[0] + x[1] - 17.25 <= 0

However, Since scipy is implemented in a way to consider the constraints with greater or equal inequalities we need to transform our constraints by multiplying it with -1:

-x[0] -x[1] + 17.25 >= 0

In [5]:
#Now we will define the inequality above as a python function.
def constraint1(x):
    return -x[0] - x[1] + 17.25

## Defining the bounds
considering that production can not have negative values and also more than the 17.25 ounce of raw material, each product is bounded between 0 and 17.25

In [6]:
b = (0.0, 17.25)
bnds = (b, b)

## Transforming Objects and Running the Opitmization Object

In [7]:
con1 = {'type': 'ineq', 'fun': constraint1}
cons = ([con1])

solution = minimize(objective,x0,method='SLSQP',bounds=bnds,constraints=cons)
x = solution.x

# show final objective
print('The total profit: ' + str(-objective(x)))

# print solution
print('Solution')
print('product 1 = ' + str(x[0]) + ' Oz')
print('product 2 = ' + str(x[1]) + ' Oz')

The total profit: 225.3749999999986
Solution
product 1 = 8.499999917604544 Oz
product 2 = 8.749999168560677 Oz


## Conclusions
The solution to our nonlinear problem suggests that by producing roughly 8.5 nunce of Product 1 and 8.75 ounce of Poduct 2, the maximum profit can be reached at 225.345 dollar.